### PR TITLE
[Qudits] Add support to cirq.channel, cirq.apply_channel, and cirq.mixture

### PR DESCRIPTION
**Pull request by cduck**
For #933. Part 3 after #1771.

- `cirq.apply_channel` checks for the `qid_shape` protocol and use properly sized numpy arrays.
- `cirq.channel` and `cirq.mixture` should just work.

Qudit features not in this PR are simulators, measurements, optimizers.

### CODE REVIEW

**Comment by cduck** [comment]
I am still adding tests for non-qubit channels and mixtures.

File: cirq/protocols/apply_channel.py [hunk 1]
```diff
@@ -324,3 +433,26 @@ def _apply_krauss_multi_qubit(krauss: Union[Tuple[Any], Sequence[Any]],
...
+        sub_args: A version of `args` with transposed and sliced views of
+            it's tensors.
+        sub_result: The result of calling an object's `_apply_channel_`
+            method on `sub_args`.  A transposed subspace of the desired
+            result.
+
+    Returns: The full result tensor after applying the channel.  Either
+        `args.target_tensor` or `args.out_buffer`.
+    """
+    return _incorporate_result_into_buffer(args.target_tensor, args.out_buffer,
```
Comment (Strilanc) [inline cirq/protocols/apply_channel.py]: Interesting that we don't have to use all the fields.

Comment (cduck) [reply cirq/protocols/apply_channel.py]: The documentation says that `_apply_channel_` isn't allowed to return either `auxiliary_buffer` as the result.

**Comment by cduck** [comment]
This incorrectly extends channel matrices.

**Pull request closed without merging**